strawberry/schema/schema_converter.py

```diff
@@ -336,6 +336,32 @@ def _check_permissions(source: Any, info: Info, kwargs: Dict[str, Any]):
                     message = getattr(permission, "message", None)
                     raise PermissionError(message)
 
+        async def _check_permissions_async(
+            source: Any, info: Info, kwargs: Dict[str, Any]
+        ):
+            for permission_class in field.permission_classes:
+                permission = permission_class()
+                has_permission: bool
+
+                if inspect.iscoroutinefunction(permission.has_permission):
+                    has_permission = await permission.has_permission(  # type: ignore
+                        source, info, **kwargs
+                    )
+                else:
+                    has_permission = permission.has_permission(  # type: ignore
+                        source, info, **kwargs
+                    )
+
+                if not has_permission:
+                    message = getattr(permission, "message", None)
+                    raise PermissionError(message)
+
+        def _has_async_permission_classes() -> bool:
```

Comment:
It would be good to hoist this to the field definition so that it only runs once. Having it here means that it will run every time the resolver is called.

Reply:
I'm undecided between calculating whether there are async permission classes in the constructor vs. having a cached property. The calculated value depends on the StrawberryField.permission_classes property which technically is mutable at the moment. I went with having a cached_property for now but I'm open for suggestions.

(we could make permission classes immutable by adding a property, or a setter that recalculates whether there are async permission classes).